File: algorithm-design/java/src/test/java/leetcode/LC022GenerateParentheses.java
package leetcode;

import org.junit.Test;

import java.util.*;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/generate-parentheses/
 */
public class LC022GenerateParentheses {
    public class Solution {
        public void generateParenthesis(String prefix, int standing, int n, Set<String> found) {
            if (n == 0 && standing == 0) {
                found.add(prefix);
                return;
            }
            // open
            if (n > 0) {
                generateParenthesis(prefix + "(", standing + 1, n - 1, found);
            }
            // close
            if (standing > 0) {
                generateParenthesis(prefix + ")", standing - 1, n, found);
            }
        }

        public List<String> generateParenthesis(int n) {
            if (n == 0) return new ArrayList<>();
            Set<String> found = new TreeSet<>();
            generateParenthesis("(", 1, n - 1, found);
            List<String> result = new ArrayList<>();
            result.addAll(found);
            return result;
        }
    }

    private void test(String[] expected, int n) throws Exception {
        List<String> result = new Solution().generateParenthesis(n);
        Collections.sort(result);
        Arrays.sort(expected);
        for (int i = 0; i < Math.max(result.size(), expected.length); i++) {
            assertEquals(expected[i], result.get(i));
        }
    }

    @Test
    public void test_0() throws Exception {
        String[] expected = {};
        test(expected, 0);
    }

    @Test
    public void test_1() throws Exception {
        String[] expected = {"()"};
        test(expected, 1);
    }

    @Test
    public void test_2() throws Exception {
        String[] expected = {"()()", "(())"};
        test(expected, 2);
    }

    @Test
    public void test_3() throws Exception {
        String[] expected = {"((()))", "(()())", "(())()", "()(())", "()()()"};
        test(expected, 3);
    }

    @Test
    public void test_4() throws Exception {
        String[] expected = {
                "(((())))", "((()()))", "((())())", "((()))()", "(()(()))", "(()()())", "(()())()", "(())(())",
                "(())()()", "()((()))", "()(()())", "()(())()", "()()(())", "()()()()"};
        test(expected, 4);
    }
}

File: algorithms/code/leetcode/lc022_generate_parentheses/lc022_generate_parentheses.rs
// https://leetcode.com/problems/generate-parentheses/

pub struct Solution;

impl Solution {
    fn generate(prefix: &str, standing: i32, n: i32, found: &mut Vec<String>) {
        if n == 0 && standing == 0 {
            found.push(prefix.to_string());
            return;
        }
        // open
        if n > 0 {
            let new_prefix = format!("{}(", prefix);
            Self::generate(&new_prefix, standing + 1, n - 1, found)
        }
        // close
        if standing > 0 {
            let new_prefix = format!("{})", prefix);
            Self::generate(&new_prefix, standing - 1, n, found)
        }
    }

    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        let mut found = Vec::new();
        if n != 0 {
            Self::generate("(", 1, n - 1, &mut found);
        }
        found
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::cmp::max;

    fn check(mut expected: Vec<String>, n: i32) {
        let mut result = Solution::generate_parenthesis(n);
        result.sort();
        expected.sort();
        for i in 0..max(result.len(), expected.len()) {
            assert_eq!(result.get(i), Some(&expected[i]));
        }
    }

    #[test]
    fn test_0() {
        let expected = Vec::new();
        check(expected, 0);
    }

    #[test]
    fn test_1() {
        let expected = vec!["()".to_string()];
        check(expected, 1);
    }

    #[test]
    fn test_2() {
        let expected = vec!["()()".to_string(), "(())".to_string()];
        check(expected, 2);
    }

    #[test]
    fn test_3() {
        let expected = vec![
            "((()))".to_string(),
            "(()())".to_string(),
            "(())()".to_string(),
            "()(())".to_string(),
            "()()()".to_string(),
        ];
        check(expected, 3);
    }

    #[test]
    fn test_4() {
        let expected = vec![
            "(((())))".to_string(),
            "((()()))".to_string(),
            "((())())".to_string(),
            "((()))()".to_string(),
            "(()(()))".to_string(),
            "(()()())".to_string(),
            "(()())()".to_string(),
            "(())(())".to_string(),
            "(())()()".to_string(),
            "()((()))".to_string(),
            "()(()())".to_string(),
            "()(())()".to_string(),
            "()()(())".to_string(),
            "()()()()".to_string(),
        ];
        check(expected, 4);
    }
}
